<robot name="myfirst" xmlns:xacro="http://wiki.ros.org/wiki/xacro">

<xacro:property name="base_length" value="2"/>
<xacro:property name="base_width" value="2"/>
<xacro:property name="base_height" value="2"/>
<xacro:property name="plat_radius" value="2"/>
<xacro:property name="plat_length" value="0.1"/>
<xacro:property name="cyl_length" value="1"/>
<xacro:property name="stem_length" value="1"/>
<xacro:property name="wheel_thickness" value="0.3"/>
<xacro:property name="wheel_radius" value="0.5"/>



  <!--base link & cyl-link-->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
      <material name="customized_color_9">
        <color rgba="0.8 0.0 0.5 1.0"/>
      </material>
    </visual>
  </link>

  <link name="cyl_link">
    <visual>
      <geometry>
        <cylinder radius="0.2" length="${cyl_length}"/>
      </geometry>
      <material name="customized_color_8">
        <color rgba="0.0 0.0 0.0 1.0"/>
      </material>
    </visual>
  </link>

  <joint name="base_cyl_joint" type="fixed">
         <origin xyz="0 0 ${base_height/2 + cyl_length/2}" rpy="0 0 0"/>
         <parent link="base_link"/>
         <child link="cyl_link"/>
  </joint>
<!--platform & cyl-link-->
  <link name="platform">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="${plat_radius}" length="${plat_length}"/>
      </geometry>
      <material name="customized_color1">
        <color rgba="0.0 0.0 0.5 1.0"/>
      </material>
    </visual>
  </link>

  <joint name="cyl_platform_joint" type="revolute">
         <origin xyz="0 0 ${cyl_length/2+plat_length/2}" rpy="0 0 0"/>
         <parent link="cyl_link"/>
         <child link="platform"/>
         <limit effort="30" velocity="1.0" lower="-1.0472" upper="1.0472" />
  </joint>
<!--base & stem-link-->
    <link name="stem">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.2" length="${stem_length}"/>
      </geometry>
      <material name="customized_color2">
        <color rgba="1.0 0.5 0.5 1.0"/>
      </material>
    </visual>
  </link>

  <joint name="base_stem_joint" type="revolute">
         <origin xyz="0 0 ${-base_length/2-stem_length/2}" rpy="0 0 0"/>
         <axis xyz="0.0 0.0 1.0"/>
         <parent link="base_link"/>
         <child link="stem"/>
         <limit effort="30" velocity="1.0" lower="-1.0472" upper="1.0472" />
  </joint>
<!--wheel & stem-link-->
    <link name="wheel">
    <visual>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
      </geometry>
      <material name="customized_color">
        <color rgba="0.8 0.5 0.5 1.0"/>
      </material>
    </visual>
  </link>

  <joint name="stem_wheel_joint" type="continuous">
         <origin xyz="0 0 ${-(stem_length)/2-wheel_radius}" rpy="0 0 0"/>
         <axis xyz="0.0 1.0 0.0"/>
         <parent link="stem"/>
         <child link="wheel"/>
  </joint>

</robot>

